File: packages/ai-setup-ts/src/generators/mode.ts
import type { GeneratedFile, Generator, GeneratorConfig, PromptQuestion } from './types.js'

function toSlug(value: string): string {
  return value
    .trim()
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, '-')
    .replace(/^-+|-+$/g, '')
}

function toTitleCase(value: string): string {
  return value
    .split(/[-_\s]+/)
    .filter(Boolean)
    .map((part) => part.charAt(0).toUpperCase() + part.slice(1))
    .join(' ')
}

export class ModeGenerator implements Generator {
  readonly type = 'mode' as const

  getPromptQuestions(): PromptQuestion[] {
    return [
      {
        key: 'description',
        label: 'Mode skill description',
        type: 'text',
        required: false,
      },
    ]
  }

  async generate(config: GeneratorConfig): Promise<GeneratedFile[]> {
    const slug = toSlug(config.name)
    const title = toTitleCase(slug || config.name)
    const description = String(config.answers?.description ?? config.description ?? `Behavioral operating mode for ${title.toLowerCase()} execution.`)

    const content = `---
kind: mode-skill
name: ${slug || 'new-mode'}
description: ${description}
behavior:
  - keep work aligned to the active plan
  - surface risks before high-cost execution
  - prefer deterministic, auditable steps
allowed_tools:
  - Read
  - Grep
  - Glob
  - Edit
  - Write
approval_policy: normal
model_hint: opus
---

# ${title} Mode Skill

Use this skill to modify how the active agent behaves while preserving the base role.

- define when to ask for confirmation
- clarify autonomy expectations
- document trade-offs and stopping conditions
`

    return [
      {
        path: `.ai/orchestration/skills/modes/${slug || 'new-mode'}.md`,
        content,
      },
    ]
  }
}

File: archive/runtime-2026-06/packages/cli/internal/generator/mode.go
package generator

import (
	"fmt"

	"github.com/rluisb/lazyai/packages/cli/internal/types"
)

// ModeGenerator generates mode skill files.
// Ported from src/generators/mode.ts.
type ModeGenerator struct{}

func (g *ModeGenerator) Type() types.ArtifactType { return types.ArtifactTypeMode }

func (g *ModeGenerator) GetPromptQuestions() []PromptQuestion {
	return []PromptQuestion{
		{
			Key:      "description",
			Label:    "Mode skill description",
			Type:     "text",
			Required: false,
		},
	}
}

func (g *ModeGenerator) Generate(config GeneratorConfig) ([]GeneratedFile, error) {
	slug := ToSlug(config.Name)
	title := ToTitleCase(slug)
	if title == "" {
		title = config.Name
	}

	description := getAnswer(config.Answers, "description", "")
	if description == "" {
		description = config.Description
	}
	if description == "" {
		description = fmt.Sprintf("Behavioral operating mode for %s execution.", stringsToLower(title))
	}

	modeSlug := slug
	if modeSlug == "" {
		modeSlug = "new-mode"
	}

	content := fmt.Sprintf(`---
kind: mode-skill
name: %s
description: %s
behavior:
  - keep work aligned to the active plan
  - surface risks before high-cost execution
  - prefer deterministic, auditable steps
allowed_tools:
  - Read
  - Grep
  - Glob
  - Edit
  - Write
approval_policy: normal
model_hint: opus
---

# %s Mode Skill

Use this skill to modify how the active agent behaves while preserving the base role.

- define when to ask for confirmation
- clarify autonomy expectations
- document trade-offs and stopping conditions
`,
		modeSlug,
		description,
		title,
	)

	return []GeneratedFile{
		{
			Path:    fmt.Sprintf(".ai/orchestration/skills/modes/%s.md", modeSlug),
			Content: content,
		},
	}, nil
}
